src/VisualStudio/Core/Def/Implementation/CodeLens/CodeLensCallbackListener.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Remote;
+using Microsoft.VisualStudio.Language.CodeLens;
+using Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem;
+using Microsoft.VisualStudio.LanguageServices.Remote;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Settings;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeLens
+{
+    /// <summary>
+    /// This is used by new codelens API to get extra data from VS if it is needed.
+    /// </summary>
+    [Export(typeof(ICodeLensCallbackListener))]
+    internal class CodeLensCallbackListener : ICodeLensCallbackListener
```

Comment:
@lineralpha how this callback match my codelens provider to this callback? is it like aggregated?

Reply:
right now there is no match between callback listener and the codelens provider since the codelens service does not play a role in the callback once it hand off the json rpc channel to the callback requester. And the callback service doesn't specify which listener object it sends to callback to. The JsonRpc finds the matching listener and the method through the json rpc method name specified in the callback request. The first matching listener will receive the callback and can respond with a result.

Reply:
shouldn't there be a way to indicate which callback one wants to talk to?

Reply:
Well, you can use JsonRpcMethodAttribute to decorate your listener's methods so they do not collide with any methods from other listeners. e.g.

[JsonRpcMethod("roslyncodelensoop:GetHostId")]
public HostId GetHostId(…)
{
}